monai/metrics/metric.py

```diff
@@ -0,0 +1,199 @@
+# Copyright 2020 - 2021 MONAI Consortium
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#     http://www.apache.org/licenses/LICENSE-2.0
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from abc import ABC, abstractmethod
+from typing import Any, List, Optional
+
+import torch
+
+from monai.config import TensorOrList
+from monai.utils import evenly_divisible_all_gather
+
+
+class Metric(ABC):
+    """
+    Base class of all Metrics inerface.
+    `__call__` is designed to execute metric computation.
+
+    """
+
+    @abstractmethod
+    def __call__(self, *args: Any, **kwds: Any):
+        """
+        API to execute the metric computation.
+
+        """
+        raise NotImplementedError(f"Subclass {self.__class__.__name__} must implement this method.")
+
+
+class IterationMetric(Metric):
+    """
+    Base class of Metrics interface for computation on a batch of tensors, usually the data of 1 iteration.
+    `__call__` is supposed to compute independent logic for several samples of `y_pred` and `y`(optional).
+    Ususally, subclass only needs to implement the `_compute_tensor` function for computation process.
+
+    """
+
+    def __call__(self, y_pred: TensorOrList, y: Optional[TensorOrList] = None):  # type: ignore
+        """
+        Execute basic computation for model prediction and ground truth.
+        It can support  both `list of channel-first Tensor` and `batch-first Tensor`.
+        And users can execute on every batch of data, then accumulate the results, or
+        accumulate the original `y_pred` and `y`, then execute on the accumulated data.
+
+        Args:
+            y_pred: the model prediction data to compute, must be a list of `channel-first` Tensor
+                or a `batch-first` Tensor.
+            y: the ground truth to compute, must be a list of `channel-first` Tensor
+                or a `batch-first` Tensor.
+
+        """
+        ret: TensorOrList
+        if isinstance(y_pred, (list, tuple)) or isinstance(y, (list, tuple)):
+            # if y_pred or y is a list of channel-first data, add batch dim and compute metric
+            ret_: List[torch.Tensor] = self._compute_list(y_pred, y)
+            # concat the list of results
```

Comment:
Hi @wyli ,

The concat logic is to guarantee the output shape of ret is `BCHW[D]`, otherwise it's `list of batch-first` which is against our principle that data in metrics should be `BCHW[D]` or `list of channel-first`.
You are right, we should do it the `_compute_list()` func, let me move it.

Thanks.

Reply:
Moved to `_compute_list()` in the latest commit.
Thanks.